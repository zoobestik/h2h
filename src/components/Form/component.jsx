import { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { serialize } from 'app/lib/form';

export default class Form extends PureComponent {
    static propTypes = {
        crc: PropTypes.string,
        onSubmit: PropTypes.func,
        children: PropTypes.node,
    };

    constructor(props) {
        super(props);
        this.onSubmit = this.onSubmit.bind(this);
    }

    onSubmit(e) {
        const onSubmit = this.props.onSubmit;

        if (onSubmit) {
            e.preventDefault();
            onSubmit(serialize(e.target), e);
        }
    }

    render() {
        const { crc, children, ...props } = this.props;

        return (
            <form method="post" { ...props } onSubmit={ this.onSubmit }>
                <input type="hidden" name="crc" value={ crc || '' }/>
                { children }
            </form>
        );
    }
}
