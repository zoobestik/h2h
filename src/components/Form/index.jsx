import { inject, observer } from 'mobx-react';
import Form from './component';

const stores2props = ({ auth }) => ({
    crc: auth.crc,
});

const FormSmart = props => <Form { ...props }/>;

export default inject(stores2props)(
    observer(FormSmart)
);
