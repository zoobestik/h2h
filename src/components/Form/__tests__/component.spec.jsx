/* eslint-env jest */
import renderer from 'react-test-renderer';
import Form from '../component';

describe('Form', () => {
    it('default', () => {
        const form = renderer.create(<Form action="/url" crc="123">Some content</Form>);
        expect(form.toJSON()).toMatchSnapshot();
    });

    it('with @className', () => {
        const form = renderer.create(<Form className="some-random-class"/>);
        expect(form.toJSON()).toMatchSnapshot();
    });

    describe('check @onSubmit', () => {
        it('with callback', () => {
            const onSubmit = jest.fn();
            const preventDefault = jest.fn();

            const form = mount(<Form action="/url/" crc="123" onSubmit={ onSubmit }>
                <input name="name1" type="text" defaultValue="value1"/>
                <textarea name="name2" defaultValue="value2"/>
            </Form>);

            form.simulate('submit', { preventDefault });

            expect(preventDefault.mocks.calls.length).toBe(1);
            expect(onSubmit.mocks.calls.length).toBe(1);
            expect(onSubmit.mocks.calls[0][0]).toEqual({
                url: '/url/',
                method: 'post',
                data: {
                    crc: '123',
                    name1: 'value1',
                    name2: 'value2',
                },
            });
        });

        it('without callback', () => {
            const preventDefault = jest.fn();
            const form = mount(<Form/>);

            form.simulate('submit', { preventDefault });

            expect(preventDefault.mock.calls.length).toBe(1);
        });
    });
});
