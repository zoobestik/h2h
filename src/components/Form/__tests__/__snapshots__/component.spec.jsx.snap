// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form default 1`] = `
<form
  action="/url"
  method="post"
  onSubmit={[Function]}
>
  <input
    name="crc"
    type="hidden"
    value="123"
  />
  Some content
</form>
`;

exports[`Form with @className 1`] = `
<form
  className="some-random-class"
  method="post"
  onSubmit={[Function]}
>
  <input
    name="crc"
    type="hidden"
    value={undefined}
  />
</form>
`;
